Source/core/Sync.cpp

```diff
@@ -782,6 +784,188 @@ namespace Core {
         return (nResult);
     }
 
+    //----------------------------------------------------------------------------
+    //----------------------------------------------------------------------------
+    // SharedSemaphore class
+    //----------------------------------------------------------------------------
+    //----------------------------------------------------------------------------
+
+    //----------------------------------------------------------------------------
+    // CONSTRUCTOR & DESTRUCTOR
+    //----------------------------------------------------------------------------
+
+#ifdef __WINDOWS__
+    SharedSemaphore::SharedSemaphore(const TCHAR sourceName[], const uint32_t initCount, const uint32_t maxCount)
+        : _semaphore(::CreateSemaphore(nullptr, initCount, maxCount, sourceName))
+    {
+        ASSERT(initCount <= 1);
+        ASSERT(maxCount <= 1);
+        ASSERT(initCount <= maxCount);
+    }
+#else
+    SharedSemaphore::SharedSemaphore(void *storage, const uint32_t initCount, const uint32_t maxCount)
+        : _semaphore(storage)
+    {
+        ASSERT(storage != nullptr);
+        ASSERT(initCount <= 1);
+        ASSERT(maxCount <= 1);
+        ASSERT(initCount <= maxCount);
+
+        if (initCount != 0 && maxCount != 0) {
+            VARIABLE_IS_NOT_USED int result =  sem_init(static_cast<sem_t*>(_semaphore), 1, initCount); 
+            ASSERT(result != -1);
+        }
+    }
+#endif
+
+    SharedSemaphore::~SharedSemaphore()
+    {
+#ifdef __WINDOWS__
+        if (_semaphore != nullptr) {
+            ::CloseHandle(_semaphore);
+        }
+#else
+        sem_destroy(static_cast<sem_t*>(_semaphore));
+#endif
+    }
+
+    size_t SharedSemaphore::Size()
+    {
+#ifdef __WINDOWS__
```

Comment:
Why would you ASSERT here? Also on window the Semaphore does have a certain size. the size of a HANDLE I guess ?

Reply:
Exactly, this has no useful meaning on Windows and if you think you need to use it on Windows, you're probably doing something wrong, hence I asked to assert here.

Reply:
@nxtum  / @sebaszm I am not religious about t, but at the end this now does leak implementation details through the interface. Seems that the static size_t Size() const operator is leaking the detai that the Size() call is *not* valid sometimes.
At the end, on windows it can not go wrong, even if you would assume that you have to allocate memory, if you than would like to use the constructor (you have to use the constructor that passes in storage space) you get a compilation error cause it does not exist (you even find the issue earlier than running it)...

@sebaszm agree that the meaning on windows is useless but disagree that you can use it incorrectly cause than you would get a compiler error..

So I have a slight preference to *not* show the implementation details and suggest to make it statis, see other remark on hiding sem_t stuff..

Reply:
The code that needs to use Size(), ie. code that will use shared memory and not a name will have to be flagged #ifndef WINDOWS anyway, so I think we could remove the Size() method on Windows completely.

Reply:
That would for me also be a viable option... Than I will run  into this at compile time and not runtime..